File: platform/extension-sdk/node/src/sql-database.ts
import { createDefaultCredential, type TokenCredentialLike } from "./credential.js";

export interface SqlDatabaseClientOptions {
  server: string;
  database: string;
  credential?: TokenCredentialLike;
  connectionTimeoutMs?: number;
  requestTimeoutMs?: number;
}

export interface SqlDatabaseClient {
  query<T = Record<string, unknown>>(sql: string, parameters?: Record<string, unknown>): Promise<T[]>;
  close(): Promise<void>;
}

export function createEntraSqlConnectionString(server: string, database: string): string {
  if (!server || server.trim().length === 0) {
    throw new Error("Azure SQL server host is required.");
  }

  if (!database || database.trim().length === 0) {
    throw new Error("Azure SQL database name is required.");
  }

  return [
    `Server=tcp:${server.trim()},1433`,
    `Initial Catalog=${database.trim()}`,
    "Authentication=Active Directory Default",
    "MultipleActiveResultSets=False",
    "Encrypt=True",
    "TrustServerCertificate=False",
    "Connection Timeout=30"
  ].join(";");
}

export async function createSqlDatabaseClient(options: SqlDatabaseClientOptions): Promise<SqlDatabaseClient> {
  const credential = options.credential ?? await createDefaultCredential();
  const token = await credential.getToken("https://database.windows.net/.default");
  const accessToken = getAccessTokenValue(token);

  const mssql = await import("mssql");
  const pool = new mssql.ConnectionPool({
    server: options.server,
    database: options.database,
    authentication: {
      type: "azure-active-directory-access-token",
      options: {
        token: accessToken
      }
    },
    connectionTimeout: options.connectionTimeoutMs,
    requestTimeout: options.requestTimeoutMs,
    options: {
      encrypt: true,
      trustServerCertificate: false
    }
  });

  const connection = await pool.connect();

  return {
    async query<T = Record<string, unknown>>(sql: string, parameters: Record<string, unknown> = {}): Promise<T[]> {
      const request = connection.request();

      for (const [name, value] of Object.entries(parameters)) {
        request.input(name, value);
      }

      const result = await request.query(sql);
      return result.recordset as T[];
    },

    async close(): Promise<void> {
      await connection.close();
    }
  };
}

function getAccessTokenValue(token: unknown): string {
  if (typeof token === "object" && token !== null && "token" in token) {
    const value = (token as { token?: unknown }).token;
    if (typeof value === "string" && value.length > 0) {
      return value;
    }
  }

  throw new Error("Unable to acquire Azure SQL access token.");
}

File: platform/extension-sdk/python/src/aks_platform_sdk/sql_database.py
from dataclasses import dataclass
from typing import Any

from .credential import create_default_credential


def create_entra_sql_connection_string(server: str, database: str) -> str:
    if not server.strip():
        raise ValueError("Azure SQL server host is required.")
    if not database.strip():
        raise ValueError("Azure SQL database name is required.")

    return (
        "Driver={ODBC Driver 18 for SQL Server};"
        f"Server=tcp:{server.strip()},1433;"
        f"Database={database.strip()};"
        "Encrypt=yes;"
        "TrustServerCertificate=no;"
        "Connection Timeout=30;"
    )


@dataclass
class SqlDatabaseClient:
    server: str
    database: str
    credential: Any | None = None

    def __post_init__(self) -> None:
        import pyodbc

        credential = self.credential or create_default_credential()
        token = credential.get_token("https://database.windows.net/.default").token
        self._connection = pyodbc.connect(
            create_entra_sql_connection_string(self.server, self.database),
            attrs_before={1256: bytes(token, "utf-16-le")},
        )

    def query(self, sql: str, parameters: tuple[Any, ...] = ()) -> list[dict[str, Any]]:
        cursor = self._connection.cursor()
        cursor.execute(sql, parameters)
        columns = [column[0] for column in cursor.description or []]
        return [dict(zip(columns, row)) for row in cursor.fetchall()]

    def close(self) -> None:
        self._connection.close()
